### ISSUE TITLE
MTU Issue..

### ISSUE THREAD

**Issue by Leozaki** [EHfive/einat-ebpf#20]
**Describe the bug**
LAN Port MTU may be limited to 1182

**Expected behavior**
ping -s 1462 1.1.1.1 work..

**Configuration**
Command-line options:

```
einat --config /etc/einet.toml
```

Configuration file:

```toml
[defaults]
ipv4_local_rule_pref = 200
ipv6_local_rule_pref = 200
ipv4_hairpin_rule_pref = 100
ipv6_hairpin_rule_pref = 100
ipv4_hairpin_table_id = 4787
ipv6_hairpin_table_id = 4787

# For ports not in specified ranges, einat would passthourgh NAT if the traffic
# is on interface's external address. You should exclude ports of services (
# e.g. SSH, HTTP server) serving on NAT host's external address and expecting
# inbound initiaed traffic from NAT port ranges specified here.
tcp_ranges = ["10000-59999"]
udp_ranges = ["10000-59999"]

# Combined ICMP query ID ranges, must include `icmp_in_ranges` and `icmp_out_ranges`.
icmp_ranges = ["0-65535"]
# Inbound ICMP query ID ranges
icmp_in_ranges = ["0-9999"]
# Outbound ICMP query ID ranges
icmp_out_ranges = ["1000-65535"]

[[interfaces]]
if_name = "ppp0"
nat44 = true
nat66 = false
ipv4_hairpin_route.internal_if_names = ["lo","enp1s0"]

# NAT records lifetimes, see <https://datatracker.ietf.org/doc/html/rfc6146#section-4> .
# See available time units in <https://github.com/fundu-rs/fundu/blob/fundu-v2.0.0/README.md#time-units> .
timeout_fragment = "2s"
timeout_pkt_min = "1m"
timeout_pkt_default = "5m"
timeout_tcp_trans = "4m"
timeout_tcp_est = "124m"

# Hairpin IP protocols. You can also add "icmp" however it would be equivalent
# to send packet back to sender due to "Endpoint-Independent Mapping" behavior
# we have and ICMP does not distinguish between source query ID and destination
# query ID, which is not very useful thus not included by default.
ip_protocols = ["tcp", "udp"]
# Defaults to `defaults.ipv4_local_rule_pref`.
ip_rule_pref = 200
# Defaults to `defaults.ipv4_hairpin_table_id`.
table_id = 4787

```

**Target Machine (please complete the following information):**

-   Architecture: [e.g. x86-64, aarch64] aarch64
-   Linux distribution: [e.g. Arch Linux, OpenWrt] Debian 12
-   Kernel version: [e.g. 5.15, 6.7.1] 6.1.0-26-arm64
-   einat version: [e.g. 0.1.1] 0.1.2
**Additional context**
Add any other context about the problem here, e.g. network interface information, firewall(iptables/nftables) configuration.

**Comment by EHfive** [comment]
Might be duplicate of #19.

>Describe the bug
>LAN Port MTU may be limited to 1182
>
>Expected behavior
>ping -s 1462 1.1.1.1 work..

Also this tells nothing, please elaborate your problem.

**Comment by EHfive** [comment]
> i tested at router.. it don't have middlebox

Show proofs, describe your processes. 

Post your results of `traceroute`s and `ping -M do -s <size> <remote>`. And if you have any iptables/nftables rules other than masquerading, remove them.

Also try using remotes other than 1.1.1.1, which could be censored if you are under GFW.

**Comment by EHfive** [comment]
I want to know the MTU from the `traceroute` MTU discover result, so you should not specify the 
packetlen `1250` in arguments.

> Try figuring out the real MTU with traceroute --mtu <remote> for ICMP and traceroute --mtu --udp <remote> for UDP.

Please **do** what I said above.

>  inet 114.x peer 100.100.0.1/32 scope global ppp0

Also try ping your PPPoE peer, if that works for packet less than 1492 bytes then there is indeed a MTU clamping somewhere.

> if i stop einat-bpf it may got full hops...

I can't got full hops till `218.2.2.2` from my end, neither with Netfilter masquerade nor einat.

Also please stop editing your comments except for typos or styling, try post it at once if possible, or just post a new one.

**Comment by EHfive** [comment]
This should be fixed with 9286f98e6ab9fe1ee58aff999adf511d30149163.

The IP header was not pulled from non-linear skb causing the whole packet do be dropped by this condition https://github.com/EHfive/einat-ebpf/blame/bd64ad38a05167c0c47b7031584eaf329861346b/src/bpf/einat.bpf.c#L1705-L1708

Previously I assumed the first IP header or Ethernet header would always be in linear section of skb thus can directly accessed, which is obviously wrong as shows by this case. So thanks for the report.

You can try snapshot builds from https://github.com/EHfive/einat-ebpf/actions/runs/11609057491. Reopen if the issue persists.
